### PR TITLE
Allow compiling ocamldoc with -principal

### PR DESCRIPTION
**Pull request by garrigue**
Due to a bug in the detection of principality for inherited methods called through self, one could not compile `ocamldoc` with `-principal`.

This PR fixes this bug, and turns on `-principal` for the `ocamldoc` and `tools` subdirectories.

### CODE REVIEW

File: Changes
Comment (trefis) [inline Changes]: That looks like a mistake.

**Comment by trefis** [comment]
I suggest that @lpw25 should review that one, given his [recent work on classes](https://github.com/ocaml/ocaml/pull/8516).

File: Changes
Comment (garrigue) [reply Changes]: Indeed, bad resolution of conflict, this is why the diff looked strange.

**Comment by garrigue** [comment]
This also solves issue #6922

**Comment by garrigue** [comment]
This may be just a stop-gap until we can merge his more extensive changes, but being able to compile ocamldoc with -principal is a clear improvement :-)

**Comment by garrigue** [comment]
This is a very simple fix, so I would like a short review, even if this might be subsumed by a bigger change.

**Comment by lpw25** [comment]
I haven't tested but my changes do indeed seem to fix this issue. Your code looks like it will do what you want, but I'm not sure it is the most natural place for the fix -- I would have thought putting something in the `Pcf_inherit` case of `class_field_aux` would make more sense.

**Comment by garrigue** [comment]
I first thought that way too, but it happens to be shorter to do it in the structure, and this ensures that we don't forget anything. And this code will disappear anyway...

**Review by lpw25 — APPROVED** [review]
Good to merge

**Comment by garrigue** [comment]
Thanks

**Pull request merged**
